subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java

```diff
@@ -74,6 +77,64 @@
  * </p>
  *
+ * <h2 id="configuration-cache">Configuration Cache</h2>
+ * <p>
+ * The <a href="https://docs.gradle.org/current/userguide/configuration_cache.html">Configuration Cache</a> handles providers in a special way.
+ * When a provider is discovered to be part of the cached build configuration (e.g. when it is a task input), one of two things happen:
+ * </p>
+ * <ul>
+ *     <li><b>Cache by value</b>.
+ *     The value of the provider is computed and stored. This typically happens if the provider is derived from purely configuration-time data.
+ *     </li>
+ *     <li><b>Cache by computation</b>. The provider is stored as a lazy computation.
+ *     For example, if the provider is a result of {@link #map(Transformer)}, then the origin provider (on which {@code map} was called) and
+ *     the {@code Transformer} (including its captured variables) are stored. Note that the origin provider may still be cached by value.
+ *     After loading from cache, the value is only computed if requested (e.g. with {@link #get()}), as usual.
+ *     This typically happens when the value of the provider is only available at execution time (e.g. derived from a task output)
+ *     or if it is sourced from the environment (e.g. a system property or an environment variable).
+ *     </li>
+ * </ul>
+ * <p>
+ * The Configuration Cache operates under the assumption that most of the stored providers will be used at execution time,
+ * so it attempts to cache values of as many providers as possible without introducing more build configuration inputs.
+ * However, the exact definition of which provider to cache is intentionally left underspecified.
+ * For example, even a provider derived from a system property provider may be cached by value if that system property is already an input.
+ * </p>
+ * <p>
+ * Following the rules below ensures that your Providers remain configuration-cache compatible regardless of applied optimizations.
+ * </p>
+ * <ol>
+ *     <li>Providers returned by {@link ProviderFactory#provider(Callable)} are always computed at configuration time.
+ *     The {@code Callable} may capture arbitrary data types and freely call configuration-time-only APIs.
+ *     This is the preferred way to bridge non-lazy and lazy APIs, like passing {@link Project#getVersion()} into task's input {@link Property}.
+ *     </li>
+ *     <li>Any value returned (provided) by the provider must conform to the
+ *     <a href="https://docs.gradle.org/current/userguide/configuration_cache_requirements.html#config_cache:requirements">configuration cache requirements</a>.
+ *     This requirement may not be enforced when the value is an intermediate result and the Configuration Cache runs the computation chain to cache the end result.
+ *     </li>
+ *     <li>Any value captured by the computation of the provider (e.g. a variable captured by the {@code Transformer} lambda supplied to {@link #map(Transformer)})
```

Comment:
💭  This rule also has an exception when it comes to `ProviderFactory.provider(Callable)`. We call this out in the rest of the list items but not here.